src/encoding.ts

```diff
@@ -0,0 +1,379 @@
+	 *
+	 * @param data
+	 * Any object that is either a string or can be coerced into a string
+	 */
+	static encode(data: any): ByteBuffer {
+		let buffer: ByteBuffer;
+
+		[ data, buffer ] = normalizeEncodingArgs(data, Ascii);
+
+		for (let i = 0, length = data.length; i < length; i++) {
+			buffer[i] = data.charCodeAt(i);
+		}
+
+		return buffer;
+	}
+
+	/**
+	 * Decode an Ascii encoded byte buffer into a string
+	 *
+	 * @param data
+	 * The byte buffer to be decoded
+	 */
+	static decode(data: ByteBuffer): string {
+		if (!validateDecodingArgs(data)) {
+			return '';
+		}
+
+		let decoded = '';
+
+		for (let i = 0, length = data.length; i < length; i++) {
+			decoded += String.fromCharCode(data[i]);
+		}
+
+		return decoded;
+	}
+}
+
+/**
+ * Utf8 codec that provides facilities for encoding a string (or string coerced object)
+ * into an Utf8 encoded byte buffer as well as decoding an Utf8 encoded byte buffer
+ * into a string.
+ * Inspired by the work of: https://github.com/mathiasbynens/utf8.js
+ */
+export class Utf8 {
+	/**
+	 * Encode a string or string coerced object into an Utf8 byte buffer
+	 *
+	 * @param data
+	 * Any object that is either a string or can be coerced into a string
+	 */
+	static encode(data: any): ByteBuffer {
+		let buffer: ByteBuffer;
+
+		[ data, buffer ] = normalizeEncodingArgs(data, Utf8);
+
+		let position = 0;
+
+		for (let i = 0, length = data.length; i < length; i++) {
+			let encodedChar = data.charCodeAt(i);
+
+			/**
+			 * Surrogates
+			 * http://en.wikipedia.org/wiki/Universal_Character_Set_characters
+			 */
+			if (encodedChar > 0xD800 && encodedChar < 0xDBFF) {
+				let lowSurrogate = data.charCodeAt(++i);
+				if (lowSurrogate >= 0xDC00 && lowSurrogate < 0xDFFF) {
+					encodedChar = 0x010000 + (encodedChar - 0xD800) * 0x0400 + (lowSurrogate - 0xDC00);
+				}
+				else {
+					i--;
+				}
+			}
+
+			if (encodedChar < 0x80) {
+				buffer[position++] = encodedChar;
+			}
+			else {
+				if (encodedChar < 0x800) {
+					buffer[position++] = ((encodedChar >> 0x06) & 0x1F) | 0xC0;
+				}
+				else if (encodedChar < 0x010000) {
+					if (encodedChar >= 0xD800 && encodedChar <= 0xDFFF) {
+						throw Error('Surrogate is not a scalar value');
+					}
+
+					buffer[position++] = ((encodedChar >> 0x0C) & 0x0F) | 0xE0;
+					buffer[position++] = ((encodedChar >> 0x06) & 0x3F) | 0x80;
+				}
+				else if (encodedChar < 0x200000) {
+					buffer[position++] = ((encodedChar >> 0x12) & 0x07) | 0xF0;
+					buffer[position++] = ((encodedChar >> 0x0C) & 0x3F) | 0x80;
+					buffer[position++] = ((encodedChar >> 0x06) & 0x3F) | 0x80;
+				}
+				buffer[position++] = (encodedChar & 0x3F) | 0x80;
+			}
+		}
+
+		buffer.length = position;
+
+		return buffer;
+	}
+
+	/**
+	 * Decode a Utf8 encoded byte buffer into a string
+	 *
+	 * @param data
+	 * The byte buffer to be decoded
+	 */
+	static decode(data: ByteBuffer): string {
+		if (!validateDecodingArgs(data)) {
+			return '';
+		}
+
+		let decoded = '';
```

Comment:
Good point about it not being in the style guide, but it was a conscious agreement during earlier discussions.

I don't really see the point to `let decoded: string = '';`, since you can obviously tell that `''` is already a string.

The one possible instance where I'd consider enforcing explicit typing even where unnecessary is for return types of exported functions.